contracts/FlashRollover.sol

```diff
@@ -0,0 +1,312 @@
+        IERC721 assetWrapper;
+        IRepaymentController repaymentController;
+        IOriginationController originationController;
+        ILoanCore targetLoanCore;
+        IERC721 targetBorrowerNote;
+    }
+
+    /* solhint-disable var-name-mixedcase */
+    // AAVE Contracts
+    // Variable names are in upper case to fulfill IFlashLoanReceiver interface
+    ILendingPoolAddressesProvider public immutable override ADDRESSES_PROVIDER;
+    ILendingPool public immutable override LENDING_POOL;
+
+    /* solhint-enable var-name-mixedcase */
+
+    constructor(ILendingPoolAddressesProvider _addressesProvider) {
+        ADDRESSES_PROVIDER = _addressesProvider;
+        LENDING_POOL = ILendingPool(_addressesProvider.getLendingPool());
+    }
+
+    function rolloverLoan(
+        RolloverContractParams calldata contracts,
+        uint256 loanId,
+        LoanLibrary.LoanTerms calldata newLoanTerms,
+        uint8 v,
+        bytes32 r,
+        bytes32 s
+    ) external override {
+        LoanLibrary.LoanData memory loanData = contracts.loanCore.getLoan(loanId);
+        _validateRollover(loanData, contracts.loanCore, contracts.targetLoanCore, newLoanTerms);
+
+        uint256 amountDue = loanData.terms.principal + loanData.terms.interest;
+        uint256 startBalance = IERC20(loanData.terms.payableCurrency).balanceOf(address(this));
+
+        {
+            address[] memory assets = new address[](1);
+            assets[0] = loanData.terms.payableCurrency;
+
+            uint256[] memory amounts = new uint256[](1);
+            amounts[0] = amountDue;
+
+            uint256[] memory modes = new uint256[](1);
+            modes[0] = 0;
+
+            OperationData memory opData = OperationData({
+                contracts: contracts,
+                loanId: loanId,
+                newLoanTerms: newLoanTerms,
+                v: v,
+                r: r,
+                s: s
+            });
+
+            bytes memory params = abi.encode(opData);
+
+            // Flash loan based on principal + interest
+            LENDING_POOL.flashLoan(address(this), assets, amounts, modes, address(this), params, 0);
+        }
+
+        // Should not have any funds leftover
+        require(
+            IERC20(loanData.terms.payableCurrency).balanceOf(address(this)) == startBalance,
+            "Changed balance after flash loan"
+        );
+    }
+
+    function executeOperation(
+        address[] calldata assets,
+        uint256[] calldata amounts,
+        uint256[] calldata premiums,
+        address initiator,
+        bytes calldata params
+    ) external override returns (bool) {
+        // TODO: Security check.
+        // Can an attacker use this to drain borrower funds? Feels like maybe
+
+        require(msg.sender == address(LENDING_POOL), "Unknown lender");
+        require(initiator == address(this), "Not initiator");
+        require(IERC20(assets[0]).balanceOf(address(this)) >= amounts[0], "Did not receive loan funds");
```

Comment:
This line assumes balance starts at 0. Elsewhere (line 120) doesn't have the same assumption. If the contract carries a balance then this could potentially be attacked to use the carried balance for loan repayment, not requiring a full flashLoan (probably not feasible attack as would require aave to call with weird behavior, but still)

I would just assume everywhere that the contract does not carry a balance -- i.e. line 120 should just check that contract balance is 0 before and after the flash loan

Reply:
Yep good spot. I remember I didn't quite lock down this thinking. Going to make your suggested change, then give a second more paranoid pass in a follow-up gas/security PR.